src/facade/dragonfly_connection.cc

```diff
@@ -934,34 +953,43 @@ void Connection::ConnectionFlow(FiberSocketBase* peer) {
   }
 }
 
-void Connection::DispatchCommand(bool has_more, absl::FunctionRef<void()> dispatch_sync,
-                                 absl::FunctionRef<MessageHandle()> dispatch_async) {
-  // Avoid sync dispatch if we can interleave with an ongoing async dispatch
-  bool can_dispatch_sync = !cc_->async_dispatch;
+void Connection::DispatchSingle(bool has_more, absl::FunctionRef<void()> invoke_cb,
+                                absl::FunctionRef<MessageHandle()> cmd_msg_cb) {
+  DCHECK(queue_backpressure_ == &tl_queue_backpressure_);
+  bool optimize_for_async = has_more;
 
-  // Avoid sync dispatch if we already have pending async messages or
-  // can potentially receive some (subscriptions > 0)
-  if (dispatch_q_.size() > 0 || cc_->subscriptions > 0)
-    can_dispatch_sync = false;
+  if (stats_->dispatch_queue_bytes > queue_backpressure_->pipeline_buffer_limit) {
+    fb2::NoOpLock noop;
+    queue_backpressure_->pipeline_cnd.wait(noop, [this] {
+      bool res = stats_->dispatch_queue_bytes < queue_backpressure_->pipeline_buffer_limit ||
+                 (dispatch_q_.empty() && !cc_->async_dispatch) || cc_->conn_closing;
+      return res;
+    });
+    if (cc_->conn_closing)
+      return;
```

Comment:
maybe if (optimize_for_async && (bytes > limit)), why'd we want to block because of pipeline conditons for non-pipeline mode? it could be a command like shutdown, etc

Reply:
good catch. fixed.